crates/app/src/lib.rs

```diff
@@ -337,29 +344,80 @@ pub fn retain_components(
 
 #[cfg(test)]
 mod test {
-    use spin_factors_test::build_locked_app;
+    use std::collections::HashSet;
+
+    use spin_manifest::schema::v2::{self, ComponentSpec, Trigger};
 
     use super::*;
 
+    fn locked_trigger(
+        trigger_type: String,
+        mut trigger: v2::Trigger,
+    ) -> anyhow::Result<LockedTrigger> {
+        fn reference_id(spec: v2::ComponentSpec) -> toml::Value {
+            let v2::ComponentSpec::Reference(id) = spec else {
+                unreachable!("should have already been normalized");
+            };
+            id.as_ref().into()
+        }
+
+        if let Some(id) = trigger.component.map(reference_id) {
+            trigger.config.insert("component".into(), id);
+        }
+
+        Ok(LockedTrigger {
+            id: trigger.id,
+            trigger_type,
+            trigger_config: trigger.config.try_into()?,
+        })
+    }
+
+    pub fn locked_app() -> LockedApp {
```

Comment:
Thanks - I think this is (unfortunately) the least worst option and I appreciate you persevering with it.

I do think you can simplify the implementation a bit, e.g.:

* The `triggers` collection can just be a variable rather than needing a whole function and a `spin-manifest` dependency:

```
vec![LockedTrigger {
    id: "trigger".into(),
    trigger_type: "dummy".into(),
    trigger_config: toml::from_str(r#"component = "empty""#).unwrap(),
}];
```

I think this is terser and clearer - `spin-manifest` has a lot of twiddly stuff that is useful for dev flexibility but causes only noise here.

* LockedComponent.source.content can be `Default::default()` (again, the loader cares but `retain_components` doesn't)